src/cargo/core/compiler/build_context/target_info.rs

```diff
@@ -748,3 +751,86 @@ impl RustcTargetData {
         self.target_config(kind).links_overrides.get(lib_name)
     }
 }
+
+/// Structure used to deal with Rustdoc fingerprinting
+#[derive(Debug, Serialize, Deserialize)]
+pub struct RustDocFingerprint {
+    pub rustc_vv: String,
+}
+
+impl RustDocFingerprint {
+    /// Read the `RustDocFingerprint` info from the fingerprint file.
+    fn read<'a, 'cfg>(cx: &Context<'a, 'cfg>) -> CargoResult<Self> {
+        let rustdoc_data = paths::read(&cx.files().host_root().join(".rustdoc_fingerprint.json"))?;
+        serde_json::from_str(&rustdoc_data).map_err(|e| anyhow::anyhow!("{:?}", e))
+    }
+
+    /// Write the `RustDocFingerprint` info into the fingerprint file.
+    fn write<'a, 'cfg>(&self, cx: &Context<'a, 'cfg>) -> CargoResult<()> {
+        paths::write(
+            &cx.files().host_root().join(".rustdoc_fingerprint.json"),
+            serde_json::to_string(&self)?.as_bytes(),
+        )
+    }
+
+    fn remove_doc_dirs(doc_dirs: &Vec<PathBuf>) -> CargoResult<()> {
+        let errs: Vec<CargoResult<()>> = doc_dirs
+            .iter()
+            .map(|path| paths::remove_dir_all(&path))
+            .filter(|res| res.is_err())
+            .collect();
+        if !errs.is_empty() {
+            Err(anyhow::anyhow!("Dir removal error"))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// This function checks whether the latest version of `Rustc` used to compile this
+    /// `Workspace`'s docs was the same as the one is currently being used in this `cargo doc`
+    /// call.
+    ///
+    /// In case it's not, it takes care of removig the `doc/` folder as well as overwriting
+    /// the rustdoc fingerprint info in order to guarantee that we won't end up with mixed
+    /// versions of the `js/html/css` files that `Rustc` autogenerates which do not have
+    /// any versioning.
+    pub fn check_rustdoc_fingerprint<'a, 'cfg>(cx: &Context<'a, 'cfg>) -> CargoResult<()> {
+        let actual_rustdoc_target_data = RustDocFingerprint {
+            rustc_vv: cx.bcx.rustc().verbose_version.clone(),
+        };
+
+        // Collect all of the target doc paths for which the docs need to be compiled for.
+        let doc_dirs: Vec<PathBuf> = cx
+            .compilation
+            .root_output
+            .iter()
+            .map(|(ck, _)| match ck {
+                CompileKind::Host => cx.files().host_root().to_path_buf(),
+                CompileKind::Target(t) => cx.files().host_root().join(Path::new(t.rustc_target())),
+            })
+            .map(|path| path.join("doc"))
+            .collect();
```

Comment:
Generally Cargo tried to avoid manually computing directory names, and instead always use `Layout` for computing them.  A simpler approach might be something like this:

```rust
        let doc_dirs: Vec<&Path> = cx
            .bcx
            .all_kinds
            .iter()
            .map(|kind| cx.files().layout(*kind).doc())
            .collect();
```

Reply:
I know you try to avoid it. And to use `fs` but didn't know about the `Layout`. Will take a deeper look. Thanks!! 
And thanks for the suggestion!